### PR TITLE
Support team namespaces for kubectl

### PR DESCRIPTION
**Pull request by gahill**

### CODE REVIEW

File: examples/charts/base/templates/configmap.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ apiVersion: v1
 kind: ConfigMap
 metadata:
   name: {{ .Values.name }}-config
-  namespace: {{ .Release.Namespace }}
+  namespace: {{ .Values.namespace }}
```
Comment (clux) [inline examples/charts/base/templates/configmap.yaml]: btw

```
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
```

will be:

```
    release: RELEASE-NAME
    heritage: Tiller
```
when not upgraded through helm. This is probably fine, even if it's false initially.

File: examples/shipcat.conf [hunk 2]
```diff
@@ -24,6 +24,12 @@ github:
 
 teams:
 - name: Doves
+  namespace: doves
+  owners:
+  - name: Eirik Albrigtsen
+    github: clux
+    slack: "@clux"
+- name: TheOthers
```
Comment (clux) [inline examples/shipcat.conf]: i am in both!? :laughing:

File: shipcat_definitions/src/config.rs [hunk 3]
```diff
@@ -93,6 +98,16 @@ impl Team {
...
             bail!("Every team must have a default notifications channel declared");
         }
+
+        if let Some(namespace) = &self.namespace {
+            use regex::Regex;
+            let re = Regex::new(r"^[a-z\-]{1,30}$").unwrap();
+
+            if !re.is_match(&namespace) {
+                bail!("Valid namespace are lower case alpha with dashes only.")
+            }
```
Comment (clux) [inline shipcat_definitions/src/config.rs]: we need a blacklist of namespaces from somewhere `["kube-system", "monitoring", "flux", ...]` should probably be a property on `Region`

File: shipcat_filebacked/src/manifest.rs [hunk 4]
```diff
@@ -157,15 +158,32 @@ impl ManifestSource {
...
     }
 
+    /// Coalesce the namespace from team and region definitions in Shipcat.conf
+    /// assumes the region will always have a namespace to fallback on
+    /// 
+    /// region namespace set | team exists | team namespace set | output
+    /// ---------------------+-------------+--------------------+---------------
+    /// yes                  | yes         | yes                | team namespace
+    /// yes                  | yes         | no                 | region namespace
+    /// yes                  | no          | no                 | region namespace
```
Comment (clux) [inline shipcat_filebacked/src/manifest.rs]: generally team's existence is verified by validate, but this is safe

**Comment by clux** [comment]
looks sensible. it might lack some changes:

- kong's upstream_uri typically go to `/{svc}.{ns}.svc.cluster.local` not sure if `{ns}` therein is picked up from manifest or from `Region`
- Region.namespace uses elsewhere ought to be reviewed

**Comment by clux** [comment]
so reconcile under this model just works presumably? initially duplicating deployments?

File: shipcat_definitions/src/config.rs (hunk 3 above)
Comment (gahill) [reply shipcat_definitions/src/config.rs]: Will add this

**Comment by gahill** [comment]
> looks sensible. it might lack some changes:
> 
> * kong's upstream_uri typically go to `/{svc}.{ns}.svc.cluster.local` not sure if `{ns}` therein is picked up from manifest or from `Region`
> * Region.namespace uses elsewhere ought to be reviewed

As discussed I will verify the usage of region vs manifest and standardise on manifest. Will add a flag to the region config so we can turn this on / off per cluster

File: shipcat_definitions/src/config.rs (hunk 3 above)
Comment (gahill) [reply shipcat_definitions/src/config.rs]: Should this be at a region level. Would we want to allow kube-system for some regions?

Comment (clux) [reply shipcat_definitions/src/config.rs]: probably not, kube-system is not for user services

**Comment by clux** [comment]
https://github.com/Babylonpartners/shipcat/issues/140 also worth tackling for the gc aspect of this

**Pull request closed without merging**